torchtune/modules/common_utils.py

```diff
@@ -163,3 +164,123 @@ def _register_reparametrize_state_dict_hooks(
     module._register_state_dict_hook(
         partial(hook, dtype=dtype, offload_to_cpu=offload_to_cpu)
     )
+
+
+@contextlib.contextmanager
+def use_persistent_kv_cache(model: nn.Module) -> Generator[None, None, None]:
+    """
+    This context manager temporarily enables KV-cacheing on a given model, which must already
+    already have KV-caches setup. All forward passes using the model within this context manager
+    will use KV-caches.
+
+    KV-caches will be enabled when entering the context manager, and will be reset and disabled upon exit,
+    without being deleted.
+
+    This is useful in cases where the maximum cache sequence length(s) is not expected to change,
+    and where we might wish to alternate between model calls which use KV-cacheing, and model calls
+    which do not use KV-cacheing, without the additional overhead of deleting and setting caches up
+    every time.
+
+    Args:
+        model (nn.Module): model to enable KV-cacheing for.
+
+    Yields:
+        None: Returns control to the caller with KV-caches enabled on the given model.
+
+    Raises:
+        ValueError: If the model does not have caches setup.
+    """
+    if not model.caches_are_setup():
+        raise ValueError(
+            "Model caches must be setup before calling persistent_kv_cache! "
+            "Please use model.setup_caches() to setup model caches."
+        )
+    for _, module in model.named_modules():
+        if hasattr(module, "kv_cache") and callable(module.kv_cache):
+            module.kv_cache.cache_enabled = True
+    try:
+        yield
+    finally:
+        for _, module in model.named_modules():
+            if hasattr(module, "kv_cache") and callable(module.kv_cache):
+                module.kv_cache.cache_enabled = True
+        model.reset_caches()
+
+
+@contextlib.contextmanager
+def setup_use_local_kv_cache(
```

Comment:
I prefer something like:

```python
with local_kv_cache():
	...
```

This sticks with the notation already known for context managers and (I believe) tells the user that the kv cache will be torn-down after the context is exited.